packages/schema/compose/src/resolve.ts

```diff
@@ -767,3 +773,45 @@ async function resolveLocalImports(
     }
   }
 }
+
+export function resolveEnviromentTypes(
+  typeInfo: TypeInfo,
+  envTypeName: EnvironmentType.QueryEnvType | EnvironmentType.MutationEnvType,
+  specificEnvType?: ObjectDefinition
+): void {
+  const genericEnvType = typeInfo.objectTypes.find(
+    (type) => type.type === "Env"
+  );
+  if (!genericEnvType) {
+    return;
+  }
+
+  if (!specificEnvType) {
+    genericEnvType.type = envTypeName;
+    return;
+  }
+
+  typeInfo.objectTypes = typeInfo.objectTypes.filter((type) => {
+    return type.type !== genericEnvType.type;
+  });
+
+  checkDuplicateEnvProperties(specificEnvType, genericEnvType.properties);
+  specificEnvType.properties.push(...genericEnvType.properties);
+}
+
+export function checkDuplicateEnvProperties(
+  envType: ObjectDefinition,
+  genericEnvProperties: AnyDefinition[]
+): void {
+  for (const specificProperty of envType.properties) {
+    if (
+      genericEnvProperties.find(
+        (genericProperty) => genericProperty.name === specificProperty.name
+      )
+    ) {
+      throw new Error(
+        `Type '${envType.type}' contains duplicate property '${specificProperty.name}' of type 'Env'`
+      );
+    }
+  }
+}
```

Comment:
```suggest
export function checkDuplicateEnvProperties(
  envType: ObjectDefinition,
  genericEnvProperties: AnyDefinition[]
): void {
  const genericEnvPropertiesSet = new Set(genericEnvProperties.map((genericProperty) => genericProperty.name))
  for (const specificProperty of envType.properties) {
    if (genericEnvPropertiesSet.has(specificProperty.name)) {
      throw new Error(
        `Type '${envType.type}' contains duplicate property '${specificProperty.name}' of type 'Env'`
      );
    }
  }
}
```

Reply:
Finding in array takes linear time while in set it take constant time. So, overall complexity of the function reduces from `O(m * n)` -> `O(m + n)` where m and n are length of properties  for specific and generic env types.